Patch removing support for nonexisting /dev/crypto. The patch can be removed
once upstream adopts it.

https://bugs.kde.org/show_bug.cgi?id=512291

--- valgrind-3.26.0/coregrind/m_syswrap/syswrap-solaris.c
+++ valgrind-3.26.0/coregrind/m_syswrap/syswrap-solaris.c
@@ -3264,25 +3264,6 @@ case VKI_FIOGETOWN:
       PRE_MEM_WRITE("ioctl(FIOGETOWN)", ARG3, sizeof(vki_pid_t));
       break;
 
-   /* CRYPTO */
-   case VKI_CRYPTO_GET_PROVIDER_LIST:
-      {
-         vki_crypto_get_provider_list_t *pl =
-            (vki_crypto_get_provider_list_t *) ARG3;
-         PRE_FIELD_READ("ioctl(CRYPTO_GET_PROVIDER_LIST, pl->pl_count)",
-                        pl->pl_count);
-
-         if (ML_(safe_to_deref)(pl, sizeof(*pl))) {
-            PRE_MEM_WRITE("ioctl(CRYPTO_GET_PROVIDER_LIST)", ARG3,
-                          MAX(1, pl->pl_count) *
-                          sizeof(vki_crypto_get_provider_list_t));
-         }
-         /* Save the requested count to unused ARG4 below,
-            when we know pre-handler succeeded.
-          */
-      }
-      break; 
-
    /* dtrace */
    case VKI_DTRACEHIOC_REMOVE:
       break;
@@ -3313,9 +3294,6 @@ }
    /* Be strict. */
    if (!ML_(fd_allowed)(ARG1, "ioctl", tid, False)) {
       SET_STATUS_Failure(VKI_EBADF);
-   } else if (ARG2 == VKI_CRYPTO_GET_PROVIDER_LIST) {
-      /* Save the requested count to unused ARG4 now. */
-      ARG4 = ARG3;
    }
 }
 
@@ -3513,21 +3491,6 @@ case VKI_FIOGETOWN:
       POST_MEM_WRITE(ARG3, sizeof(vki_pid_t));
       break;
 
-   /* CRYPTO */
-   case VKI_CRYPTO_GET_PROVIDER_LIST:
-      {
-         vki_crypto_get_provider_list_t *pl =
-            (vki_crypto_get_provider_list_t *) ARG3;
-
-         POST_FIELD_WRITE(pl->pl_count);
-         POST_FIELD_WRITE(pl->pl_return_value);
-
-         if ((ARG4 > 0) && (pl->pl_return_value == VKI_CRYPTO_SUCCESS))
-            POST_MEM_WRITE((Addr) pl->pl_list, pl->pl_count *
-                           sizeof(vki_crypto_provider_entry_t));
-      }
-      break;
-
    /* dtrace */
    case VKI_DTRACEHIOC_REMOVE:
    case VKI_DTRACEHIOC_ADDDOF:
--- valgrind-3.26.0/include/vki/vki-solaris.h
+++ valgrind-3.26.0/include/vki/vki-solaris.h
@@ -336,14 +336,6 @@ } vki_da_u;
 } vki_kcf_door_arg_t;
 
 
-#include <sys/crypto/ioctl.h>
-#define VKI_CRYPTO_SUCCESS CRYPTO_SUCCESS
-#define VKI_CRYPTO_GET_PROVIDER_LIST CRYPTO_GET_PROVIDER_LIST
-#define vki_crypto_provider_id_t crypto_provider_id_t
-#define vki_crypto_provider_entry_t crypto_provider_entry_t
-#define vki_crypto_get_provider_list_t crypto_get_provider_list_t
-
-
 #include <sys/dditypes.h>
 #include <sys/devinfo_impl.h>
 #define VKI_DINFOUSRLD DINFOUSRLD
--- valgrind-3.26.0/memcheck/tests/solaris/scalar_ioctl.c
+++ valgrind-3.26.0/memcheck/tests/solaris/scalar_ioctl.c
@@ -5,7 +5,6 @@ #define __EXTENSIONS__ 1
 #include "scalar.h"
 
 #include <net/if.h>
-#include <sys/crypto/ioctl.h>
 #include <sys/dditypes.h>
 #include <sys/devinfo_impl.h>
 #include <sys/dtrace.h>
@@ -459,25 +458,6 @@ GO(SYS_ioctl, "(FIOGETOWN) 3s 1m");
    SY(SYS_ioctl, x0 - 1, x0 + FIOGETOWN, x0 + 1); FAIL;
 }
 
-/* crypto */
-__attribute__((noinline))
-static void sys_ioctl_CRYPTO_GET_PROVIDER_LIST(void)
-{
-   GO(SYS_ioctl, "(CRYPTO_GET_PROVIDER_LIST) 3s 1m");
-   SY(SYS_ioctl, x0 - 1, x0 + CRYPTO_GET_PROVIDER_LIST, x0 + 1); FAIL;
-}
-
-__attribute__((noinline))
-static void sys_ioctl_CRYPTO_GET_PROVIDER_LIST_2(void)
-{
-   crypto_get_provider_list_t pl;
-
-   pl.pl_count = x0 + 1;
-
-   GO(SYS_ioctl, "(CRYPTO_GET_PROVIDER_LIST) 4s 0m");
-   SY(SYS_ioctl, x0 - 1, x0 + CRYPTO_GET_PROVIDER_LIST, &pl + x0); FAIL;
-}
-
 /* dtrace */
 __attribute__((noinline))
 static void sys_ioctl_DTRACEHIOC_REMOVE(void)
@@ -578,10 +558,6 @@ /* filio */
    sys_ioctl_FIOSETOWN();
    sys_ioctl_FIOGETOWN();
 
-   /* crypto */
-   sys_ioctl_CRYPTO_GET_PROVIDER_LIST();
-   sys_ioctl_CRYPTO_GET_PROVIDER_LIST_2();
-
    /* dtrace */
    sys_ioctl_DTRACEHIOC_REMOVE();
    sys_ioctl_DTRACEHIOC_ADDDOF();
--- valgrind-3.26.0/memcheck/tests/solaris/scalar_ioctl.stderr.exp
+++ valgrind-3.26.0/memcheck/tests/solaris/scalar_ioctl.stderr.exp
@@ -868,38 +868,6 @@ ...
  Address 0x........ is not stack'd, malloc'd or (recently) free'd
 
 ---------------------------------------------------------
- 54:               SYS_ioctl (CRYPTO_GET_PROVIDER_LIST) 3s 1m
----------------------------------------------------------
-Syscall param ioctl(fd) contains uninitialised byte(s)
-   ...
-
-Syscall param ioctl(request) contains uninitialised byte(s)
-   ...
-
-Syscall param ioctl(arg) contains uninitialised byte(s)
-   ...
-
-Syscall param ioctl(CRYPTO_GET_PROVIDER_LIST, pl->pl_count) points to unaddressable byte(s)
-   ...
- Address 0x........ is not stack'd, malloc'd or (recently) free'd
-
----------------------------------------------------------
- 54:               SYS_ioctl (CRYPTO_GET_PROVIDER_LIST) 4s 0m
----------------------------------------------------------
-Syscall param ioctl(fd) contains uninitialised byte(s)
-   ...
-
-Syscall param ioctl(request) contains uninitialised byte(s)
-   ...
-
-Syscall param ioctl(arg) contains uninitialised byte(s)
-   ...
-
-Syscall param ioctl(CRYPTO_GET_PROVIDER_LIST, pl->pl_count) points to uninitialised byte(s)
-   ...
- Address 0x........ is on thread 1's stack
-
----------------------------------------------------------
  54:               SYS_ioctl (DTRACEHIOC_REMOVE) 3s 0m
 ---------------------------------------------------------
 Syscall param ioctl(fd) contains uninitialised byte(s)
